### ISSUE TITLE
`kops replace -f "$CLUSTER_CONFIG" --force` should delete removed instancegroups

### ISSUE THREAD

**Issue by wendorf** [kubernetes/kops#4632]
------------- FEATURE REQUEST TEMPLATE --------------------

1. Describe IN DETAIL the feature/behavior/change you would like to see.

I am using `kops replace -f "$CLUSTER_CONFIG" --force` in my CI pipeline to idempotently apply a cluster config with one `kind: Cluster`, and multiple `kind: InstanceGroup`. If I modify an existing instance group, or add a new one, kops correctly detects the change and creates or modifies the actual instance group when I do a followup `kops update cluster "$CLUSTER_NAME" --yes`.

However, if I remove an instance group from my "$CLUSTER_CONFIG" and do a `replace` and `update`, the instance group remains. Additionally, when I `kops get instancegroups --name "$CLUSTER_NAME"-o yaml`, I see the instance group I attempted to remove.

2. Feel free to provide a design supporting your feature request.

It would be great if `kops replace -f "$CLUSTER_CONFIG" --force` used the "$CLUSTER_CONFIG" entirely and removed instance groups not present in the new config. I would prefer this to be the default behavior, but having a flag like `--delete-omitted-instancegroups` would work for me, too.

Thanks!

**Comment by rdrgmnzs** [comment]
I'm also getting hit by this, removing an IG from the config and doing a replace does not remove the IG.

**Comment by mcasimir** [comment]
Was also looking into that.

I'm doing `kops get k8s-cluster.example.com -o yaml > current-status.yaml` to get the previous status and then delete what is not in the `desired-status.yaml`, but is not so amazing as is quite a delicate operation and it would be great to have that officially supported.

**Comment by xulsitatirev** [comment]
Definitely +1. I am facing the exact same problem.
As I handle many `k8s` clusters, I do keep their configuration at `./clusters` config.
There clusters are generated from `toolbox`. Ten CI/CD pipeline applies the change.
I would expect it to behave like `terraform apply` on the state.

Full code: https://github.com/xulsitatirev/k8s-kops

**Comment by jonjomckay** [comment]
This is hitting us a bit too, as it's making our infrastructure automation not-quite-so-automated anymore.

**Comment by xulsitatirev** [comment]
@jonchiu I workarounded this problem by differentiation cluster provisioning and cluster upgrade.
It is enough to do not include `sshCredential` template while upgrading.

**Comment by ms4720** [comment]
bit me too in testing

**Comment by queimadus** [comment]
Same issue here.

Tried going around the problem by removing all instancegroups in the target cluster prior to running `replace --force` but since `kops delete ig` removes the definition AND the cloud resources it doesn't work out.

**Comment by ms4720** [comment]
perhaps a --force-all to force deletions as well without breaking current behavior?

**Comment by xulsitatirev** [comment]
@ms4720 Is it proper `kops` parameter?
I would rather see a rolling upgrade instead of deletion all resources at once. :-D

**Comment by ms4720** [comment]
@xulsitatirev that would work better, as the last step in the update probably.  It needs to be managed.

**Comment by mbelang** [comment]
Any progress on this? Had the same issue while testing and I think I broke something because deletion was not done properly and now in a weird state where unable to `update` nor `delete` the cluster completely. Had to manually delete some resources from AWS and the state of kops is broken...

**Comment by xulsitatirev** [comment]
@mbelang `secret` was properly overwritten to me,  but unfortunately, instance-groups are still affected.

**Comment by selslack** [comment]
/reopen
/remove-lifecycle rotten

This feature should exist.

**Comment by wendorf** [comment]
@selslack I agree. I just got tired of reopening the issue every month (especially since I'm not currently using kops). I encourage you to open a new issue. Feel free to use the body of this issue if you want.

**Comment by selslack** [comment]
@wendorf can you do `/reopen` here, please? I will try to keep this issue alive after that.

**Comment by wendorf** [comment]
Sure thing!

/reopen

**Comment by selslack** [comment]
@johngmyers hi, what does "frozen" mean in this repo's workflow?

**Comment by johngmyers** [comment]
It means it won't go stale

**Comment by olemarkus** [comment]
I imagine one would want a `sync` command that not only disruptively deletes all IGs, but:
* Updates/rolls (or force the user to do a rolling-update) any IGs with changes not marked for deletion
* Adds all new IGs mentioned in a spec file
* Drains IGs to be deleted
* Deletes IGs
* Performs validation between each steps.

The `delete ig` command terminates the entire ASG blindly and may be more disruptive than one things. Worst case, a lot of things will stop working because the remaining ASGs don't have capacity to fit all the pods.
